src/librustc_typeck/check/mod.rs

```diff
@@ -1761,6 +1761,7 @@ pub fn check_simd(tcx: TyCtxt<'_>, sp: Span, def_id: DefId) {
             match e.sty {
                 ty::Param(_) => { /* struct<T>(T, T, T, T) is ok */ }
                 _ if e.is_machine() => { /* struct(u8, u8, u8, u8) is ok */ }
+                ty::Array(ty, _c) if ty.is_machine() => { /* struct([f32; 4]) */ }
                 _ => {
                     span_err!(tcx.sess, sp, E0077,
                               "SIMD vector element type should be machine type");
```

Comment:
Now that I think about it, "machine type" is *such a weird name*, especially for something user-facing.

I think it comes from a time where `int` was a big int and `iN` was the "machine type" equivalent (but I'm not sure this ever happened). I'd rather it say "primitive scalar (integer/float/pointer)" (do we allow `bool` and `char`?).

cc @nikomatsakis

Reply:
> do we allow bool and char?

No, we do not. 

>  I'd rather it say "primitive scalar (integer/float/pointer)" 

I'd rather have a `is_simd_element_ty` check that just checks what we want to support for SIMD vectors.

Reply:
I'm talking about the error message. For the internal method call, something like `is_valid_(for-)simd_element(_ty)` is fine (name's a bikeshed anyway).

Reply:
I've improved the error messages. The internal method is still called `.is_machine`.